File: mobile/android/app/src/main/java/com/playbridge/sender/connection/ConnectionCoordinator.kt
package com.playbridge.sender.connection

import android.util.Log
import com.playbridge.sender.cast.MediaTrack
import com.playbridge.sender.cast.TvPlaybackStatus
import com.playbridge.sender.cast.TvPlayerSettings
import com.playbridge.sender.library.PlaylistEpisode
import com.playbridge.sender.library.PlaylistUiState
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.launch

/**
 * TV browser User-Agent state, as last reported by the TV: [active] is the name of the
 * selected entry (blank = default/no override), [entries] are the name→value pairs saved
 * on the TV (for reselection — see `createUserAgentJson`/`IncomingMessage.UserAgent`).
 */
data class TvUserAgentState(
    val active: String = "",
    val entries: List<Pair<String, String>> = emptyList(),
)

/**
 * Coordinates WebSocket connection state parsing, TV playback updates, and command execution.
 * Extracts message processing out of BrowserActivity for better separation of concerns.
 */
class ConnectionCoordinator(
    private val webSocketClient: WebSocketClient,
    private val scope: CoroutineScope
) {
    private val TAG = "ConnectionCoordinator"

    // TV Playback & Playlist States synced from WebSocket
    val tvActiveContext = MutableStateFlow("idle")
    val tvPlaylistState = MutableStateFlow<PlaylistUiState?>(null)
    val tvPlayback = MutableStateFlow<TvPlaybackStatus?>(null)
    val tvVideoTracks = MutableStateFlow<List<MediaTrack>>(emptyList())
    val tvAudioTracks = MutableStateFlow<List<MediaTrack>>(emptyList())
    val tvSubtitleTracks = MutableStateFlow<List<MediaTrack>>(emptyList())
    val tvPlayerSettings = MutableStateFlow(TvPlayerSettings())
    
    // Names of user scripts currently installed on the TV (for the management UI).
    val installedUserScripts = MutableStateFlow<List<String>>(emptyList())

    // TV browser User-Agent: which one is active (name, blank = default) + which custom
    // ones are saved on the TV (for the management UI).
    val tvUserAgentState = MutableStateFlow(TvUserAgentState())

    // TMDb Sync & Now Playing Metadata states
    val nowPlayingTvId = MutableStateFlow<Int?>(null)
    val nowPlayingSeason = MutableStateFlow<Int?>(null)
    val nowPlayingEpisodeStart = MutableStateFlow(1)

    init {
        startListening()
    }

    private fun startListening() {
        scope.launch {
            webSocketClient.messages.collect { message ->
                try {
                    val json = org.json.JSONObject(message)
                    when (json.optString("type")) {
                        "context" -> {
                            val active = json.optString("active", "idle")
                            tvActiveContext.value = active
                            Log.d(TAG, "TV Active Context updated: $active")
                            // Clear playlist and playback states when TV goes idle
                            if (active == "idle") {
                                clearPlayerStates()
                            }
                        }
                        "playlist_status" -> {
                            val itemsJson = json.optJSONArray("items")
                            val episodes = buildList {
                                if (itemsJson != null) {
                                    for (i in 0 until itemsJson.length()) {
                                        val o = itemsJson.optJSONObject(i) ?: continue
                                        add(
                                            PlaylistEpisode(
                                                index = o.optInt("index", i),
                                                title = o.optString("title", "Item ${i + 1}"),
                                                season = o.optInt("season", -1).takeIf { it >= 0 },
                                                episode = o.optInt("episode", -1).takeIf { it >= 0 },
                                                imdbId = o.optString("imdbId", "").ifEmpty { null },
                                                bingeGroup = o.optString("bingeGroup", "").ifEmpty { null },
                                                mediaKind = o.optString("mediaKind", "video")
                                                    .takeIf { it in setOf("video", "audio", "image") }
                                                    ?: "video",
                                            )
                                        )
                                    }
                                }
                            }
                            tvPlaylistState.value = PlaylistUiState(
                                currentIndex = json.optInt("currentIndex", 0),
                                totalCount = json.optInt("totalCount", 0),
                                items = episodes
                            )
                            Log.d(
                                TAG,
                                "TV playlist_status index=${json.optInt("currentIndex", 0)} " +
                                    "total=${json.optInt("totalCount", 0)} items=${episodes.size}",
                            )
                        }
                        "status" -> {
                            // Perf: dedupe identical ticks — the TV status cadence (~1/s)
                            // otherwise emits a distinct object per tick and recomposes
                            // every collector. Raw ms precision kept so progress bars
                            // glide instead of jumping in 1s steps.
                            val next = TvPlaybackStatus(
                                state = json.optString("state", "paused"),
                                positionMs = json.optLong("position", 0L),
                                durationMs = json.optLong("duration", 0L),
                                title = json.optString("title", "").ifEmpty { null },
                                mediaKind = json.optString("mediaKind", "video")
                                    .takeIf { it in setOf("video", "audio", "image") }
                                    ?: "video",
                            )
                            if (next != tvPlayback.value) tvPlayback.value = next
                        }
                        "tracks" -> {
                            fun parseTracks(arr: org.json.JSONArray?): List<MediaTrack> =
                                buildList {
                                    if (arr != null) {
                                        for (i in 0 until arr.length()) {
                                            val o = arr.optJSONObject(i) ?: continue
                                            add(
                                                MediaTrack(
                                                    id = o.optString("id"),
                                                    name = o.optString("name", "Track ${i + 1}"),
                                                    selected = o.optBoolean("selected", false),
                                                    type = if (o.has("type")) o.optString("type") else null
                                                )
                                            )
                                        }
                                    }
                                }
                            tvVideoTracks.value = parseTracks(json.optJSONArray("video"))
                            tvAudioTracks.value = parseTracks(json.optJSONArray("audio"))
                            tvSubtitleTracks.value = parseTracks(json.optJSONArray("subtitle"))
                            Log.d(TAG, "TV Audio/Subtitle tracks updated")
                        }
                        "user_scripts" -> {
                            val arr = json.optJSONArray("names")
                            installedUserScripts.value = buildList {
                                if (arr != null) for (i in 0 until arr.length()) add(arr.optString(i))
                            }
                            Log.d(TAG, "TV user scripts: ${installedUserScripts.value}")
                        }
                        "user_agents" -> {
                            val active = json.optString("active", "")
                            val entriesJson = json.optJSONArray("entries")
                            val entries = buildList {
                                if (entriesJson != null) {
                                    for (i in 0 until entriesJson.length()) {
                                        val o = entriesJson.optJSONObject(i) ?: continue
                                        add(o.optString("name") to o.optString("value"))
                                    }
                                }
                            }
                            tvUserAgentState.value = TvUserAgentState(active, entries)
                            Log.d(TAG, "TV user agents updated: active=$active, ${entries.size} saved")
                        }
                        "player_settings" -> {
                            tvPlayerSettings.value = TvPlayerSettings(
                                speed = json.optDouble("speed", 1.0).toFloat(),
                                scaling = json.optString("scaling", "Fit"),
                                audioBoost = json.optBoolean("audioBoost", false),
                                subtitleOffsetMs = json.optLong("subtitleOffsetMs", 0L),
                                engine = json.optString("engine", ""),
                                qualityMaxHeight = json.optInt("qualityMaxHeight", 0),
                                currentVideoHeight = json.optInt("currentVideoHeight", 0),
                                isLive = json.optBoolean("isLive", false),
                                isSeekable = json.optBoolean("isSeekable", true),
                                speedAvailable = json.optBoolean("speedAvailable", true),
                                scalingAvailable = json.optBoolean("scalingAvailable", true),
                                audioBoostAvailable = json.optBoolean("audioBoostAvailable", true),
                                qualityAvailable = json.optBoolean("qualityAvailable", false),
                            )
                            Log.d(TAG, "TV Player settings updated")
                        }
                    }
                } catch (e: Exception) {
                    Log.e(TAG, "Error parsing WebSocket message: ${e.message}", e)
                }
            }
        }
    }

    /**
     * Begin a locally-initiated playback session: record the library identity (nulls
     * for unidentified content like browser videos/phone files), clear the playback
     * snapshots left over from the previous session — the TV pushes fresh status and
     * playlist_status within a second, but the progress tracker must never pair the
     * NEW identity with the OLD position (it would instantly "watch" the new item) —
     * and flip the context to "player".
     */
    fun startLocalPlaybackSession(tmdbId: Int?, season: Int?, episodeStart: Int?) {
        nowPlayingTvId.value = tmdbId
        nowPlayingSeason.value = season
        nowPlayingEpisodeStart.value = episodeStart ?: 1
        clearPlaybackSnapshots()
        tvActiveContext.value = "player"
    }

    /** Drop the last-known status/playlist (stale once new content is being sent). */
    fun clearPlaybackSnapshots() {
        tvPlayback.value = null
        tvPlaylistState.value = null
    }

    /**
     * Mark the receiver idle locally (e.g. session ended from the phone while no native
     * target is connected to confirm it). Keeps external classes from poking
     * [tvActiveContext] directly.
     */
    fun markIdle() {
        tvActiveContext.value = "idle"
    }

    private fun clearPlayerStates() {
        tvPlaylistState.value = null
        tvPlayback.value = null
        tvVideoTracks.value = emptyList()
        tvAudioTracks.value = emptyList()
        tvSubtitleTracks.value = emptyList()
        tvPlayerSettings.value = TvPlayerSettings()
        nowPlayingTvId.value = null
        nowPlayingSeason.value = null
        nowPlayingEpisodeStart.value = 1
        Log.d(TAG, "TV Player states cleared as TV is idle")
    }
}

File: mobile/apple/PlayBridge Phone/PlayBridge Phone/Network/ConnectionCoordinator.swift
import Foundation
import Combine

/// Parses TV→phone status messages into observable now-playing state.
/// Direct port of `connection/ConnectionCoordinator.kt`. Handles the proto-defined messages
/// (context/status/playlist_status) plus the non-proto `tracks` and `player_settings`.
final class ConnectionCoordinator: ObservableObject {
    @Published var activeContext: String = "idle"   // player | browser | idle
    @Published var playback: TvPlaybackStatus?
    @Published var playlist: PlaylistUiState?
    @Published var audioTracks: [MediaTrack] = []
    @Published var subtitleTracks: [MediaTrack] = []
    @Published var playerSpeed: Float = 1.0
    @Published var playerScaling: String = "Fit"
    @Published var playerIsLive = false
    @Published var playerIsSeekable = true
    @Published var speedAvailable = false
    @Published var scalingAvailable = false
    @Published var mediaKind = "video"
    @Published var lastCommandResult: QueueCommandResult?

    /// Feed every non-handshake message here (wired to `WebSocketClient.onMessage`).
    func handle(_ text: String) {
        guard let data = text.data(using: .utf8),
              let json = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any],
              let type = json["type"] as? String else { return }

        DispatchQueue.main.async { self.dispatch(type: type, json: json) }
    }

    private func dispatch(type: String, json: [String: Any]) {
        switch type {
        case "context":
            let active = json["active"] as? String ?? "idle"
            activeContext = active
            if active == "idle" { clear() }

        case "status":
            mediaKind = json["mediaKind"] as? String ?? "video"
            playback = TvPlaybackStatus(
                state: json["state"] as? String ?? "paused",
                positionMs: int64(json["position"]),
                durationMs: int64(json["duration"]),
                title: (json["title"] as? String).flatMap { $0.isEmpty ? nil : $0 },
                playbackId: nonEmpty(json["playbackId"]),
                currentItemId: nonEmpty(json["currentItemId"])
            )

        case "playlist_status":
            let itemsJson = json["items"] as? [[String: Any]] ?? []
            var episodes: [PlaylistEpisode] = []
            for (i, o) in itemsJson.enumerated() {
                let index: Int = (o["index"] as? Int) ?? i
                let title: String = (o["title"] as? String) ?? "Item \(i + 1)"
                var season: Int? = o["season"] as? Int
                if let s = season, s < 0 { season = nil }
                var episode: Int? = o["episode"] as? Int
                if let e = episode, e < 0 { episode = nil }
                let imdb = (o["imdbId"] as? String).flatMap { $0.isEmpty ? nil : $0 }
                let tmdb = (o["tmdbId"] as? String).flatMap { $0.isEmpty ? nil : $0 }
                let binge = (o["bingeGroup"] as? String).flatMap { $0.isEmpty ? nil : $0 }
                episodes.append(PlaylistEpisode(
                    index: index, title: title, itemId: nonEmpty(o["itemId"]),
                    season: season, episode: episode, imdbId: imdb, tmdbId: tmdb,
                    bingeGroup: binge
                ))
            }
            playlist = PlaylistUiState(
                currentIndex: json["currentIndex"] as? Int ?? 0,
                totalCount: json["totalCount"] as? Int ?? 0,
                items: episodes,
                playbackId: nonEmpty(json["playbackId"]),
                currentItemId: nonEmpty(json["currentItemId"]),
                queueRevision: uint64(json["queueRevision"])
            )

        case "command_result":
            guard let requestId = json["requestId"] as? String,
                  let ok = json["ok"] as? Bool else { return }
            lastCommandResult = QueueCommandResult(
                requestId: requestId,
                ok: ok,
                error: nonEmpty(json["error"]),
                playbackId: nonEmpty(json["playbackId"]),
                queueRevision: json["queueRevision"].map { uint64($0) }
            )

        case "tracks":
            audioTracks = parseTracks(json["audio"])
            subtitleTracks = parseTracks(json["subtitle"])

        case "player_settings":
            if let speed = json["speed"] as? Double { playerSpeed = Float(speed) }
            playerScaling = json["scaling"] as? String ?? playerScaling
            playerIsLive = json["isLive"] as? Bool ?? false
            playerIsSeekable = json["isSeekable"] as? Bool ?? true
            speedAvailable = json["speedAvailable"] as? Bool ?? true
            scalingAvailable = json["scalingAvailable"] as? Bool ?? true

        default:
            break
        }
    }

    private func parseTracks(_ value: Any?) -> [MediaTrack] {
        guard let arr = value as? [[String: Any]] else { return [] }
        return arr.enumerated().map { (i, o) in
            MediaTrack(
                id: (o["id"] as? String) ?? (o["id"] as? NSNumber)?.stringValue ?? "\(i)",
                name: o["name"] as? String ?? "Track \(i + 1)",
                selected: o["selected"] as? Bool ?? false
            )
        }
    }

    private func int64(_ value: Any?) -> Int64 {
        if let n = value as? Int64 { return n }
        if let n = value as? Int { return Int64(n) }
        if let n = value as? Double { return n.isFinite ? (Int64(exactly: n.rounded(.towardZero)) ?? 0) : 0 }
        if let n = value as? NSNumber { return n.int64Value }
        return 0
    }

    private func uint64(_ value: Any?) -> UInt64 {
        if let n = value as? UInt64 { return n }
        if let n = value as? Int { return UInt64(max(0, n)) }
        if let n = value as? NSNumber { return n.uint64Value }
        return 0
    }

    private func nonEmpty(_ value: Any?) -> String? {
        (value as? String).flatMap { $0.isEmpty ? nil : $0 }
    }

    func clear() {
        playback = nil
        playlist = nil
        audioTracks = []
        subtitleTracks = []
        playerSpeed = 1.0
        playerScaling = "Fit"
        playerIsLive = false
        playerIsSeekable = true
        speedAvailable = false
        scalingAvailable = false
        mediaKind = "video"
        lastCommandResult = nil
    }
}
